src/cpp/src/block_manager.hpp

```diff
@@ -44,15 +50,79 @@ class KVCacheBlock {
     int get_references_count() const {
         return m_ref_count;
     }
+
+    size_t get_hash() const {
+        return m_hash;
+    }
+
+    size_t get_num_hashed_tokens() const {
+        return m_num_hashed_tokens;
+    }
+
+    void set_hash(size_t hash, size_t num_hashed_tokens) {
+        m_hash = hash;
+        m_num_hashed_tokens = num_hashed_tokens;
+    }
+
+    void set_timestamp(const std::chrono::time_point<std::chrono::system_clock>& timestamp) {
+        m_timestamp = timestamp;
+    }
+
+    std::chrono::time_point<std::chrono::system_clock> get_timestamp() {
+        return m_timestamp;
+    }
+};
+
+
+class Evictor {
+    std::map<size_t, KVCacheBlock::Ptr> blocks;
+public:
+    void add(size_t hash, KVCacheBlock::Ptr block) {
+        blocks[hash] = block;
+    }
+
+    static bool block_is_less(const std::pair<size_t, KVCacheBlock::Ptr>& lhs, const std::pair<size_t, KVCacheBlock::Ptr>& rhs) {
+        return lhs.second->get_timestamp() < rhs.second->get_timestamp();
+    }
+
+    KVCacheBlock::Ptr get_block(size_t hash) {
+        if (blocks.find(hash)== blocks.end())
+        {
+            return nullptr;
+        }
+        KVCacheBlock::Ptr block = blocks[hash];
+        block->set_timestamp(std::chrono::system_clock::now());
+        block->increment();
+        blocks.erase(hash);
+        return block;
+    }
+
+    KVCacheBlock::Ptr get_lru_block() {
+        if (!blocks.size()) {
+            return nullptr;
+        }
+        auto hash_block = std::min_element(std::begin(blocks), std::end(blocks), block_is_less);
```

Comment:
do you think we need to store blocks in already sorted manner?
In this case `get_lru_block` will take `O(1)`

Reply:
If we store blocks in a sorted structure like `priority_queue`, `get_lru_block()` will be `O(1)` but `get_block()` by hash will be `O(n)`, as we will need to loop over all blocks and check hashes:
https://github.com/openvinotoolkit/openvino.genai/blob/22ef6c40356edf854accca434768b8cd76aa617d/src/cpp/src/block_manager.hpp#L88

Currently with hash table `get_block()` is `O(1)`. We can probably have two structures in evictor both hash table and `priority_queue`?

Reply:
ok, we can try this optimization.

But maybe let's first perform comparison with vLLM implementation? We can measure how much time this piece of code takes averagely.